### DISCUSSION TITLE
Why does Terraform not clone my target machine properly?

### DISCUSSION THREAD

**Discussion by PinguPongu** [Telmate/terraform-provider-proxmox#1060] [Q&A] [answered]
Hi
What I want to achieve is simply to bulk clone a Windows machine, similarly how you can clone a machine inside Proxmox. 
But whenever I run my code it successfully creates a VM but it doesn't seem to base it on the clone at all, since I need to set up the OS from scratch as if it were a fresh Windows install, unlike the Proxmox functionality.
I've tried omitting the ` iso = "local:iso/SERVER_EVAL_x64FRE_en-us.iso"` part, but then it never boots properly.
What am I doing wrong, and how can I ensure the VM is based on the clone without needing a fresh OS setup?

```
resource "proxmox_vm_qemu" "WindowsTofu" {
    count        = 2
    name         = "WindowsTofu-${count.index}"
    desc         = "Windows test tofu test 1"
    vmid        = count.index + var.vm_id
    target_node = "myNode"
    agent       = 1
    clone        = "WS-template"
    full_clone  = true
    define_connection_info = false

    cores       = 4
    sockets     = 1
    memory      = 8192

    os_type     = "win11" 

    disks {
        ide {
            ide0 {
                cdrom {
                    iso = "local:iso/SERVER_EVAL_x64FRE_en-us.iso"
                }
            }
        }

        scsi {
            scsi0 {
                disk {
                    storage = "local-lvm"
                    size    = 20
                }
            }
        }
    }

    network {
        model  = "virtio"
        bridge = "vmbr0"
    }
    
    efidisk {
        efitype = "4m"
        storage = "local-lvm"
    }
}
```

**Comment by maksimsamt** [comment]
Try to remove these config lines:

```hcl
...
        ide {
            ide0 {
                cdrom {
                    iso = "local:iso/SERVER_EVAL_x64FRE_en-us.iso"
                }
            }
        }
...
```

Your template is `uefi` and the main disk is `scsi0`?
If yes, add `bios` and `boot` arguments:
```hcl
...
bios        = "ovmf"
boot        = "order=scsi0"
...
```

Probably, you should disable Agent (does it exist on Windows?):
```hcl
...
agent       = 0
...
```

Not sure regarding this, if this is correct:
```hcl
...
os_type     = "win11" 
...
```

**Comment by PinguPongu** [comment]
Hi! 
Sorry for the late response, and thank you for the response! :)
I've tried all your suggestions in all sorts of combinations.

What I have right now is the following:
```
variable "vm_id" {
  default = 700
}

resource "proxmox_vm_qemu" "WindowsTofu" {
    count       = 2
    name        = "WindowsTofu-${count.index}"
    desc        = "Windows test tofu test 1"
    vmid        = count.index + var.vm_id
    target_node = "<target-node>"
    agent       = 0
    clone       = "WS-template"
    bios        = "ovmf"
    boot        = "order=scsi0"
    full_clone  = true
    define_connection_info = false

    cores       = 4
    sockets     = 1
    memory      = 8192

    disks {

        scsi {
            scsi0 {
                disk {
                    storage = "local-lvm"
                    size    = 20
                }
            }
        }
    }

    network {
        model  = "virtio"
        bridge = "vmbr0"
    }
    
    efidisk {
        efitype = "4m"
        storage = "local-lvm"
    }
}
```

And when removing the following block:
```
        ide {
            ide0 {
                cdrom {
                    iso = "local:iso/SERVER_EVAL_x64FRE_en-us.iso"
                }
            }
        }
```

I will always get booted to a bootscreen stating the following (and never actually get booted into windows):
```
>>Start PXE over IPv4.
   Station IP address is <Station IP Address>

   Server IP address is <Server IP Address>
   NBP filename is linux26    # I don't know why it is referencing 'linux26' here
   NBP filesize is 0 Bytes
   PXE-E18: Server response timeout.
   BdsDxe: failed to load Boot0001 "UEFI PXEv4 (MAC:<MAC Address>)" from PciRoot(0x0)/Pci(0x12,0x0)/MAC(<MAC Address>,0x1)/IPv4(0.0.0.0,0.0.0.0,0x0,0)///DHCP,0.0.0.0,0.0.0.0,0.0.0.0,0.0.0.0,0.0.0.0: Not Found

>>Start PXE over IPv6.
```

Why is this simple task proving to be so hard for me?

**Comment by maksimsamt** [comment]
What will be if you add this?
```hcl
...
pxe       = false
...
```

Also, have you tried provisioning just one VM instead of bulk?

**Reply by PinguPongu** [reply]
Did that and I got the following error:
`"pxe": conflicts with clone`

Again thanks for the help.

**Reply by maksimsamt** [reply]
So, yes, this is correct, you should not use together clone and pxe:
> [Note that pxe is mutually exclusive with clone modes.](https://github.com/Telmate/terraform-provider-proxmox/blob/master/docs/resources/vm_qemu.md#top-level-block)

**Comment by maksimsamt** [comment]
And what about this?
> Your template is uefi and the main disk is scsi0?

**Reply by PinguPongu** [reply]
Yes. 
I also think it is worth noting the full Hardware config of my Template:
![image](https://github.com/user-attachments/assets/4be14c8e-c480-4846-8dab-4ad35e66abcd)

**Reply by maksimsamt** [reply, marked as answer]
1) In your template there is `ide0` disk, not `scsi0`, so, you have two options:
- change your template to use `scsi0` disk
- change your Terraform config to use `ide0` disk

2) Your template disk is 64G, in your Terraform config only 20G. Target VM should have the same or larger disk (>=64G, not 20G)

3) Is it correct, that your template disk is TPM encrypted? If so, I doubt that it is supported by the current terraform provider.

4) As for `CPU [x86-64-v2-AES]`. 
By default terraform provider sets it to `host`. If you need any other option, you need to add it to the Terraform configuration, for example

```hcl
...
cpu =  "x86-64-v2-AES"
...
```

**Reply by PinguPongu** [reply]
I can't believe it. It actually worked!
Thank you so much man! You really saved me. Yeah the issue was the TPM and the obvious fact that I was using sci0 in my terraform. 
Thank you so so so much

**Comment by PinguPongu** [comment]
I even tried a simplistic code but, got booted in to a broken machine with the same boot message as before.
```
resource "proxmox_vm_qemu" "WindowsClone" {
    count       = 1
    name        = "WindowsClone"
    clone       = "WS-template"
    node       = <node>
    full_clone  = true
}
```
I don't get it. What am I missing? I feel like this should be so simple.

**Reply by maksimsamt** [reply]
> I feel like this should be so simple.

Yes, it can be simple if you do it right and if current terraform provider supports it!

**Comment by l0wl3vel** [comment]
I am running into a very similar problem. I have a heavily modified template VM and using it as template in this provider would lead to me having to basically redefine it explicitly, which I would rather avoid, since I would have to make changes at two places now. This is not, as @maksimsamt alleged, simple.

If you clone a VM using the UI or the [API](https://pve.proxmox.com/pve-docs/api-viewer/#/nodes/{node}/qemu/{vmid}/clone) the semantics are completly different. You just inherit every aspect of the template.

I think there is a proper solution here. We just would need to adjust thesources of truth Terraform uses on resource creation when cloning. Those are:

### TF Provider Specified
Non-Null default values are specified in the provider for every parameter that is not overridden by the user. This one is currently used and the provider explicitly defines the parameters of the VM. It makes sense that there is a default here from a usability POV but in my opinion the default provided by the system should take preference if it is provided.

### Deferred System Default
These are values that are resulting in a (known after apply ) message when used in the output. The underlying system (Proxmox) defines it an Terraform adopts whatever the value is into the state.

To solve our problem we would need to make all our variables use whatever the Deferred System Default value is when cloning the resource. It would require to not set the default value as the desired value, unless the user specified it explicitly.  We should stick to how the [API](https://pve.proxmox.com/pve-docs/api-viewer/#/nodes/{node}/qemu) semantics work on how to set defaults.
